keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,1032 @@
+
+[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
+- Additional tests are in Testgrid and linked in KEP
+
+#### GA
+
+- N examples of real-world usage
+- N installs
+- More rigorous forms of testing—e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+-->
+
+#### Alpha
+
+- Feature disabled by default.
+- Unit and integration tests completed and passed.
+- Additional tests are in Testgrid and linked in KEP
+
+#### Beta
+
+- Feature enabled by default.
+- Gather feedback from developers and out-of-tree plugins.
+- Benchmark tests passed, and there is no performance degradation.
+- Update documents to reflect the changes.
+
+#### GA
+
+- Fix all reported bugs.
+- Feature enabled and cannot be disabled. All feature gate guarded logic are removed.
+- Update documents to reflect the changes.
+
+### Upgrade / Downgrade Strategy
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to maintain previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to make use of the enhancement?
+-->
+
+- Upgrade
+  - While the feature gate is enabled, `.spec.schedulingPaused` can be mutated by users.
+  - While the feature gate is enabled and `.spec.schedulingPaused` is not set, a default value will
+  be given to behave the same like it's specified by users.
+- Downgrade
+  - Previously configured values will be ignored.
+
```

Comment:
We need to mentioned that this is an opt-in feature since users need to explicitly set the field in the pod spec. If the feature is disabled the field is preserved if it was already set in the persisted pod object, otherwise it is silently dropped.

Reply:
I suppose the downgrade means "downgrading to a version that this feature is not supported", but I will reword to make it clear.

Reply:
my comment is more about how we handle updates to the field when the feature gate is on/off, not only k8s version downgrades.

Reply:
Hmm... the downgrade case is based on feature gate being off, so do you mean to emphasize that the field is preserved until it gets updated for the first time after downgrading?

Reply:
Correct, so when the feature gate is on, then we accept setting the field, when the feature is off, we always drop the field on create, but on update there are two cases for update: 1) if it was set before, then we don't touch (we don't drop, but we also don't update it); 2) if the field was never set before, we drop it. However, since we don't allow adding to the field, then case 2 is actually no relevant to this feature.

Reply:
Makes sense to me.